contributors/design-proposals/termination-reason.md

```diff
@@ -0,0 +1,317 @@
+progress.)
+1. As a user I want to be able to send a termination reason to my Pods when I 
+use the Kubernetes API or kubectl to gracefully delete a Pod.
+1. As a Reliability Engineer for an application running on a Kubernetes cluster,
+I want to configure Pods to communicate the reason that they were terminated via 
+a HTTP GET request so that I can configure monitoring and alerting based on 
+historic trends.
+
+## Assumptions
+
+1. Kubelet will always invoke the pre-stop handler prior to sending a `TERM` 
+signal to the entry point of a container.
+1. pre-stop handlers will not contain complicated or long running business logic. 
+The logic associated with container cleanup will be executed by the container's
+entry point during the container's configured termination grace period.
+1. Termination reason is a general string that can be associated with an 
+ObjectMeta at that time of deletion. Currently, it only has semantic meaning 
+with respect to the graceful deletion of individual Pods. We do not, at this
+time, propose to implement controller assisted propagation of termination 
+reasons from one API object to another (e.g Setting a termination reason for 
+a ReplicaSet which the controller will then propagate to its Pods). Such
+considerations will be deferred to future work.
+
+## Requirements
+
+1. Termination reason must be available to a container at the time its pre-stop 
+handler is invoked. In this way, applications can perform any actions
+necessary to configure the container's entry point prior to the entry point 
+receiving the `TERM` signal.
+1. The delivery mechanism for termination reason must be consumable by, and 
+ compatible with, existing HTTP and command based pre-stop handlers.
+1. Any headers, variables, or parameters used to communicate a termination 
+reason must not be hard coded, and the receiving container must be able to 
+override the defaults.
+
+## API Objects
+
+Termination reason is implemented as a generic string that can be associated 
+with the metadata of any object that supports graceful termination. Initially, 
+the reason will only be actionable with respect to Pods' containers, but this 
+generic delivery mechanism can be extended to other use cases.
+
+DeleteOptions is modified to carry a reason.
+
+```golang
+type DeleteOptions struct {
+	// Other fields omitted for brevity.
+	
+	// Reason indicates the reason for the deletion of an API Object that 
+	// undergo graceful termination upon deletion.
+	Reason string `json:"reason,omitempty"`
+}
+```
+
+The ObjectMeta struct is modified to carry the deletion reason via 
+an Object's metadata.
+
+```golang
+type ObjectMeta {
+	// Other fields omitted for brevity.
+	
+    // DeletionReason indicates the reason for the deletion of an API Object. 
+    // Its purpose is to provide an extra generic signal to watchers of API 
+    // Objects during the graceful termination process. 
+	DeletionReason string `json:"reason,omitempty"`
```

Comment:
So here's an extra question. 

If I need to use this (abuse this) for additional metadata... why wouldn't I just put that in an annotation?